benchmarks/benchmark-dense-solve.C

```diff
@@ -0,0 +1,181 @@
+
+#include "linbox/linbox-config.h"
+#include <iostream>
+
+#include <givaro/modular.h>
+#include "linbox/util/args-parser.h"
+#include "linbox/matrix/sparse-matrix.h"
+#include "linbox/solutions/solve.h"
+#include "linbox/util/matrix-stream.h"
+#include "linbox/solutions/methods.h"
+
+#ifdef _DEBUG
+#define _BENCHMARKS_DEBUG_
+#endif
+
+using namespace LinBox;
+typedef Givaro::ZRing<Givaro::Integer> Ints;
+
+int main (int argc, char **argv)
+{
+    Givaro::Integer q = -1 ;
+    size_t nbiter = 3 ;
+    size_t n = 500 ;
+    size_t bits = 10;
+//     size_t p = 0;
+
+    Argument as[] = {
+        { 'i', "-i R", "Set number of repetitions.",       TYPE_INT , &nbiter },
+        { 'q', "-q Q", "Set the field characteristic (-1 for rationals).", TYPE_INTEGER , &q },
+        { 'n', "-n N", "Set the matrix dimension.",      TYPE_INT , &n },
+        { 'b', "-b B", "bit size", TYPE_INT , &bits },
+//         { 'p', "-p P", "0 for sequential, 1 for 2D iterative, 2 for 2D rec, 3 for 2D rec adaptive, 4 for 3D rec in-place, 5 for 3D rec, 6 for 3D rec adaptive.", TYPE_INT , &p },
+        END_OF_ARGUMENTS
+    };
+
+    LinBox::parseArguments(argc,argv,as);
+
+    bool ModComp = false;
+    if (q > 0) ModComp = true;
+
+    
+    Timer chrono;
+    std::vector<std::pair<double,double>> timebits(nbiter);
+    for(size_t iter=0; iter<nbiter; ++iter) {
+
+        if (ModComp) {
+            
+            typedef Givaro::Modular<double> Field;
+            Field F(q);
+            Field::RandIter G(F);
+	
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "Setting A ... " << std::endl;
+#endif
+            chrono.start();
+            DenseMatrix<Field> A(F,n,n);
+            PAR_BLOCK { FFLAS::pfrand(F,G, n,n,A.getPointer(),n); }
+            chrono.stop();
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "... A is " << A.rowdim() << " by " << A.coldim() << ", " << chrono << std::endl;
+            if (A.rowdim() <= 20 && A.coldim() <= 20) A.write(std::clog <<"A:=",Tag::FileFormat::Maple) << ';' << std::endl;
+#endif            
+            DenseVector<Field> X(F, A.coldim()),B(F, A.rowdim());
+            for(auto it=B.begin(); it != B.end(); ++it)
+                if (drand48() <0.5)
+                    F.assign(*it,F.mOne);
+                else
+                    F.assign(*it,F.one);
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "B is [";
+            for(const auto& it: B)
+                F.write(std::clog, it) << ' ';
+            std::clog << ']' << std::endl;
+#endif
+
+                // DenseElimination
+            chrono.start();
+            solve (X, A, B, Method::DenseElimination());
+            chrono.stop();
+
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "(DenseElimination) Solution is [";
+            for(const auto& it: X)
+                F.write(std::clog, it) << ' ';
+            std::clog << ']' << std::endl;
+#endif
+            timebits[iter].second=Givaro::logtwo(GIVMAX(X.front(), 1));
+        } else {
```

Comment:
Avoid duplicating the same code for the 2 instances: Modular<double> and ZRing<Integer>. Maybe write the same code in a `run_with_field` method templated by the Field type and have `main` call this function twice in a if then else block.

Reply:
The code is not the same, randiter and solve have different API

Reply:
Using rational interface (X = BlasVector<Rational>) should do it, or do I miss another difference?

Reply:
Hum then X elements do not have the same interface either (I want the bitsize of the denominator)

Reply:
OK, I concede to factorize the code with template specializations